### PR TITLE
ensure consistent code formatting for the c++ code

### LINKED ISSUE #201: Set travis to check if the code in pull request is formatted correctly
Travis should check the files in the pull request (only those files) and if any of the files do not follow the code style then the build should be marked as broken.
Preferably, this should be done before doing the compiling&testing phase (travis staging? )

Get the list of files from pull request 194:
```
curl -s "https://api.github.com/repos/espressopp/espressopp/pulls/194/files" | grep '"filename":' | while read a; do echo ">${a}"; done
```
or with `git diff` 
```
git diff --name-only $TRAVIS_COMMIT_RANGE
```

cc: #197

### PR DESCRIPTION
**Pull request by XzzX**
Using `clang-format` to ensure consistent formatting. Run `format.sh` in the root folder to format the whole code base.

closes #202 
closes #201

### CODE REVIEW

**Comment by junghans** [comment]
I would integrate it in CMake instead of using a shell script.

https://github.com/ECP-copa/Cabana/blob/master/CMakeLists.txt#L205

**Comment by XzzX** [comment]
> I would integrate it in CMake instead of using a shell script.
> 
> https://github.com/ECP-copa/Cabana/blob/master/CMakeLists.txt#L205

Make targets are only available in the build folder. Why should I go to the build folder to format my source folder? I consider the wild mix of source and build folders in `espressopp` bad practice.

**Comment by junghans** [comment]
True, but we need a way to check the clang-format version, otherwise it will be a constant back and forth between version.

plus CMake support Windows ;-)

File: .github/workflows/validate.yml [hunk 1]
```diff
@@ -129,3 +129,14 @@ jobs:
       - name: Image digest
         run: echo ${{ steps.docker_build.outputs.digest }}
 
+  formatting:
+    runs-on: ubuntu-latest
```
Comment (junghans) [inline .github/workflows/validate.yml:133]: I would run that in one of our containers as `ubuntu-latest` has 3 different clang versions.

**Comment by junghans** [comment]
We should also run pep8 on all the python!

**Comment by XzzX** [comment]
> True, but we need a way to check the clang-format version, otherwise it will be a constant back and forth between version.

Hmm, enforcing a specific `clang-format` version seems awkward. Which version would you use? Default in Ubuntu 18, 20, fedora? We also have to force the users to have the same version... Or we hope that they all produce the same results. I formatted with `clang-format-12`, seems to be consistent with what github actions provides.

> 
> plus CMake support Windows ;-)

True, have you ever tested `espressopp` on Windows? If not I have serious doubts it will compile. My experience with `waLBerla` (we support Windows) is that MSVC is tricky at times. ;) And there is only support for OpenMP 2.0...

**Comment by XzzX** [comment]
> We should also run pep8 on all the python!

On step after the other :) Let us see what the others say in today's meeting.

File: .github/workflows/validate.yml (hunk 1 above)
Comment (XzzX) [reply .github/workflows/validate.yml:133]: Which one?

**Comment by XzzX** [comment]
Different versions of `clang-format` seem to be an issue.
https://stackoverflow.com/questions/59395507/how-to-maintain-a-clang-format-file-for-different-clang-format-versions
Ideas? Force a certain version?

**Comment by junghans** [comment]
In VOTCA we use the version that comes with fedora:latest. Ubuntu would be ok as well, that might have the advantage that there is only a new lts release every couple of years.

**Comment by XzzX** [comment]
> In VOTCA we use the version that comes with fedora:latest. Ubuntu would be ok as well, that might have the advantage that there is only a new lts release every couple of years.

Ok, we can try it like this. If troubles arise we can look at it again.
I am not really troubled with updates of the `clang-format` version within a distribution, more with users having multiple different installations...

**Comment by XzzX** [comment]
Since this change influences everyone, who should give their ok?

**Comment by junghans** [comment]
In VOTCA we also have an action that allow formatting of a PR (see https://github.com/votca/tools/blob/master/.github/workflows/format.yml), we could to the same for espressopp to help developers in case they cannot get the right version of `clang-format`.

**Comment by XzzX** [comment]
> In VOTCA we also have an action that allow formatting of a PR (see https://github.com/votca/tools/blob/master/.github/workflows/format.yml), we could to the same for espressopp to help developers in case they cannot get the right version of `clang-format`.

This is an option, yes, at the cost of adding additional commits.

**Comment by XzzX** [comment]
> In VOTCA we also have an action that allow formatting of a PR (see https://github.com/votca/tools/blob/master/.github/workflows/format.yml), we could to the same for espressopp to help developers in case they cannot get the right version of `clang-format`.

I see another problem with this approach. This does not stop multiple changes in PRs due to formatting that will get reverted afterwards. This makes reviewing more complicated.

**Comment by junghans** [comment]
> > In VOTCA we also have an action that allow formatting of a PR (see https://github.com/votca/tools/blob/master/.github/workflows/format.yml), we could to the same for espressopp to help developers in case they cannot get the right version of `clang-format`.
> 
> I see another problem with this approach. This does not stop multiple changes in PRs due to formatting that will get reverted afterwards. This makes reviewing more complicated.

Yeah it only makes sense to the formatting once at the end or simply do a squash merge.

**Comment by junghans** [comment]
I just realized there is a `format` target already in CMake.

**Comment by XzzX** [comment]
> I just realized there is a `format` target already in CMake.
> 
> https://github.com/espressopp/espressopp/blob/5df088dcc353cad0c472e0ac1965e4dfb2cb6acd/src/CMakeLists.txt#L96-L105

This PR would remove this target.

**Comment by XzzX** [comment]
My current suggestion:

1) Convert bash script to Python. This should allow potential Windows users to also run the script.
2) Run format check per commit but not as required. Reviewers can then manually recommend running the script to users.

I am a little bit hesitant to include a format&commit bot as I am afraid it will make change history less useful.

@junghans @jnvance Opinions?

**Comment by junghans** [comment]
From my experience on other projects even with a pre-commit hook basically everybody forgets to format and then everything is reformated in the last commit. There is a clang tool to run formatting on each commit, but that can generate a lot of conflicts.

I thought about this quite a bit but I don't have a good solution.

Looking at the most recent PRs our history is pretty useless already, so one more formatting commit won't make it much worse. Maybe we should just squash merge PRs by default, then formatting commits would not matter.

**Comment by XzzX** [comment]
> From my experience on other projects even with a pre-commit hook basically everybody forgets to format and then everything is reformated in the last commit. 

Use an IDE that supports format on save :)

> Maybe we should just squash merge PRs by default, then formatting commits would not matter.

That is a very good suggestion. I like it. 

If I understand the VOTCA bot you can call him via the comment function and he produces a format commit? Are the custom actions and votca bot user still required to implement this functionality?

**Comment by junghans** [comment]
The only reason we need a bot account is that only a commit from a PAT can trigger actions and usually you want to run the CI again after formatting.

**Review by junghans — APPROVED** [review]

File: .github/workflows/format-command.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+name: format-command
+on:
+  repository_dispatch:
+    types: [format-command]
+jobs:
+  example:
```
Comment (junghans) [inline .github/workflows/format-command.yml:6]: Might be not a great name.

**Review by junghans — CHANGES_REQUESTED** [review]

File: .github/workflows/format-command.yml (hunk 2 above)
Comment (XzzX) [reply .github/workflows/format-command.yml]: changed to format

**Review by junghans — APPROVED** [review]

**Pull request merged**